bug: npm-kubo fails to install on windows

windows has a history of issues with symlinks. i don't know why this is popping up now, but there are multiple reports of this not working, and I was able to reproduce locally on a windows machine as well.


### Kubo Releaser failure

```
> ./kuboreleaser release --version v0.23.0 update-ipfs-desktop
GPG_KEY is not set. The key should be base64 encoded. Please enter the key:: 
GPG_PASSPHRASE is not set. Please enter a secret value: 
INFO[2023-10-05 18:00:37] Checking the status of the action...         
INFO[2023-10-05 18:00:37] I'm going to check if the PR that updates Kubo in IPFS Desktop has been created already. 
INFO[2023-10-05 18:00:37] The action is not complete yet, continuing... 
WARN[2023-10-05 18:00:37] PR for https://github.com/ipfs/ipfs-desktop/tree/kubo-v0.23.0 not found (the action is not complete) 
INFO[2023-10-05 18:00:37] Running the action...                        
INFO[2023-10-05 18:00:37] I'm going to create a PR that updates Kubo in IPFS Desktop. 
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
npm ERR! code 1
npm ERR! path /tmp/kuboreleaser2456904201/node_modules/kubo
npm ERR! command failed
npm ERR! command sh -c node src/post-install.js
npm ERR! https://dist.ipfs.tech/kubo/versions
npm ERR! Downloading https://dist.ipfs.tech/kubo/v0.23.0/kubo_v0.23.0_linux-amd64.tar.gz to /root/.cache/npm-kubo
npm ERR! Downloaded https://dist.ipfs.tech/kubo/v0.23.0/kubo_v0.23.0_linux-amd64.tar.gz
npm ERR! Downloading kubo_v0.23.0_linux-amd64.tar.gz.sha512
npm ERR! Downloaded kubo_v0.23.0_linux-amd64.tar.gz.sha512
npm ERR! Verifying kubo_v0.23.0_linux-amd64.tar.gz.sha512
npm ERR! OK (1cbfc09f4b6ed13481b4035bce8414824ed019afefc866378daa1f3f1cbf4e91306f8525e600074eaac392b527a53fd9a662406183abdc4bd80bb37dc29a7472)
npm ERR! Unpacked /tmp/kuboreleaser2456904201/node_modules/kubo
npm ERR! Linking /tmp/kuboreleaser2456904201/node_modules/kubo/kubo/ipfs to /tmp/kuboreleaser2456904201/node_modules/kubo/bin/ipfs
npm ERR! Error: ipfs binary failed: Error: spawnSync /tmp/kuboreleaser2456904201/node_modules/kubo/bin/ipfs ENOENT
npm ERR!     at link (/tmp/kuboreleaser2456904201/node_modules/kubo/src/download.js:220:11)
npm ERR!     at module.exports (/tmp/kuboreleaser2456904201/node_modules/kubo/src/download.js:248:10)
npm ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-10-05T18_00_43_346Z-debug-0.log
exit status 1
```

### local npm failure
```
PS C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo> npm i
npm WARN deprecated trim@0.0.1: Use String.prototype.trim() instead

> kubo@0.23.0 postinstall
> node src/post-install.js

https://dist.ipfs.tech/kubo/versions
Downloading https://dist.ipfs.tech/kubo/v0.23.0/kubo_v0.23.0_windows-amd64.zip to C:\Users\sgtpo\AppData\Local\npm-kubo\Cache       
Downloaded https://dist.ipfs.tech/kubo/v0.23.0/kubo_v0.23.0_windows-amd64.zip
Downloading kubo_v0.23.0_windows-amd64.zip.sha512
Downloaded kubo_v0.23.0_windows-amd64.zip.sha512
Verifying kubo_v0.23.0_windows-amd64.zip.sha512
OK (3c6c0c6a90012bc37c8afdc380e48bda4b0413218aaccf72fa246e10915c1c2d04ab3e2211fbdede97a0bca75b00f2be0f6ce9e320829b45dcc23f14b93c8a08)
Unpacked C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo
Linking C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\kubo\ipfs.exe to C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\bin\ipfs.exe
Error: EPERM: operation not permitted, symlink 'C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\kubo\ipfs.exe' -> 'C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\bin\ipfs.exe'
    at Object.symlinkSync (node:fs:1816:3)
    at link (C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\src\download.js:207:6)
    at module.exports (C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo\src\download.js:248:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  errno: -4048,
  syscall: 'symlink',
  code: 'EPERM',
  path: 'C:\\Users\\sgtpo\\code\\work\\protocol.ai\\ipfs\\npm-kubo\\kubo\\ipfs.exe',
  dest: 'C:\\Users\\sgtpo\\code\\work\\protocol.ai\\ipfs\\npm-kubo\\bin\\ipfs.exe'
}
npm ERR! code 1
npm ERR! path C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo  
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node src/post-install.js

npm ERR! A complete log of this run can be found in: C:\Users\sgtpo\AppData\Local\npm-cache\_logs\2023-10-06T17_28_19_739Z-debug-0.log
PS C:\Users\sgtpo\code\work\protocol.ai\ipfs\npm-kubo> 
```

### github CI failure

```
Run npm ci --prefer-offline --no-audit --progress=false --cache D:\a\ipfs-desktop\ipfs-desktop/.cache/npm
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v[8](https://github.com/ipfs/ipfs-desktop/actions/runs/6422809980/job/17471972505#step:9:9).dev/blog/math-random for details.
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! errno -4058
npm ERR! enoent ENOENT: no such file or directory, lstat
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 
npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! errno -4058
npm ERR! enoent ENOENT: no such file or directory, lstat
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 
npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! errno -4058
npm ERR! enoent ENOENT: no such file or directory, lstat
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in: D:\a\ipfs-desktop\ipfs-desktop\.cache\npm\_logs\2023-[10](https://github.com/ipfs/ipfs-desktop/actions/runs/6422809980/job/17471972505#step:9:11)-06T[16](https://github.com/ipfs/ipfs-desktop/actions/runs/6422809980/job/17471972505#step:9:17)_04_[21](https://github.com/ipfs/ipfs-desktop/actions/runs/6422809980/job/17471972505#step:9:22)_504Z-debug-0.log
```


### References
* slack reports: https://filecoinproject.slack.com/archives/C04M8232QRW/p1696528939482799
* github CI windows failures: https://github.com/ipfs/ipfs-desktop/actions/runs/6422809980/job/17471972505

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -9,15 +9,15 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
       - name: Check for new kubo version
         id: check
         uses: ./.github/actions/check-for-kubo-release
       - name: Set up node
         if: steps.check.outputs.publish == 'true'
-        uses: actions/setup-node@v3
+        uses: actions/setup-node@v4
         with:
-          node-version: 14.x
+          node-version: 20.x
       - name: Install
         if: steps.check.outputs.publish == 'true'
         run: npm install

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -8,12 +8,16 @@ on:
 
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [windows-latest, macos-latest, linux-latest]
+        node-version: [20.x]
     steps:
-    - uses: actions/checkout@v3
-    - uses: actions/setup-node@v3
+    - uses: actions/checkout@v4
+    - uses: actions/setup-node@v4
       with:
-        node-version: 14.x
+        node-version: ${{ matrix.node-version }}
     - run: npm install
     - uses: gozala/typescript-error-reporter-action@v1.0.9
     - run: npm run build --if-present

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -11,7 +11,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [windows-latest, macos-latest, linux-latest]
+        os: [windows-latest, macos-latest, ubuntu-latest]
         node-version: [20.x]
     steps:
     - uses: actions/checkout@v4

--- a/README.md
+++ b/README.md
@@ -49,6 +49,8 @@ This module downloads Kubo (go-ipfs) binaries from https://dist.ipfs.tech into y
 
 It will download the kubo version that matches the npm version of this module. So depending on `kubo@0.23.0` will install `kubo v0.23.0` for your current system architecture, in to your project at `node_modules/kubo/kubo/ipfs` and additional symlink to it at `node_modules/kubo/bin/ipfs`.
 
+On Windows, `ipfs.exe` file is used, and if the symlink can't be created under a regular user, a copy of `ipfs.exe` is created instead.
+
 After downloading you can find out the path of the installed binary by calling the `path` function exported by this module:
 
 ```javascript

--- a/src/download.js
+++ b/src/download.js
@@ -192,6 +192,9 @@ async function link ({ depBin, version }) {
   let localBin = path.resolve(path.join(__dirname, '..', 'bin', 'ipfs'))
 
   if (isWin) {
+    if (fs.existsSync(localBin)) {
+      fs.unlinkSync(localBin)
+    }
     localBin += '.exe'
   }
 

--- a/src/download.js
+++ b/src/download.js
@@ -204,7 +204,23 @@ async function link ({ depBin, version }) {
   }
 
   console.info('Linking', depBin, 'to', localBin)
-  fs.symlinkSync(depBin, localBin)
+  try {
+    fs.symlinkSync(depBin, localBin)
+  } catch (err) {
+    // Try to recover when creating symlink on modern Windows fails (https://github.com/ipfs/npm-kubo/issues/68)
+    if (isWin && typeof err === 'object' && err !== null && 'code' in err && err.code === 'EPERM') {
+      console.info('Symlink creation failed due to insufficient privileges. Attempting to copy file instead...')
+      try {
+        fs.copyFileSync(depBin, localBin)
+        console.info('Copying', depBin, 'to', localBin)
+      } catch (copyErr) {
+        console.error('File copy also failed:', copyErr)
+        throw copyErr
+      }
+    } else {
+      throw err
+    }
+  }
 
   if (isWin) {
     // On Windows, update the shortcut file to use the .exe

--- a/test/download.js
+++ b/test/download.js
@@ -3,7 +3,7 @@
 const test = require('tape-promise').default(require('tape'))
 const fs = require('fs-extra')
 const os = require('os')
-const path= require('path')
+const path = require('path')
 const { download, downloadAndUpdateBin } = require('../src/download')
 const { path: detectLocation } = require('../')
 const clean = require('./fixtures/clean')
@@ -72,7 +72,7 @@ test('Ensure calling download function manually with static values works', async
     platform: 'darwin',
     arch: 'arm64',
     distUrl: 'https://dist.ipfs.tech',
-    installPath: tempDir,
+    installPath: tempDir
   })
   console.log(kuboPath)
   const stats = await fs.stat(kuboPath)

--- a/test/path.js
+++ b/test/path.js
@@ -4,6 +4,10 @@ var path = require('path')
 var cp = require('child_process')
 var ipfs = path.join(__dirname, '..', 'bin', 'ipfs')
 
+if (process.platform === 'win32') {
+  ipfs += '.exe'
+}
+
 test('ensure ipfs bin path exists', function (t) {
   t.plan(4)
   fs.stat(ipfs, function (err, stats) {

